### PR TITLE
CTF Refine Analyze

### PR DESCRIPTION
**Pull request by rmarabini**
This pull request implements analyze  in CTF-refine.

In particular displays up to 5 items:

1)  A plot per micrograph showing for each particle the difference between the defocus assigned to the micrograph and the defocus assigned to the particle (this is equivalent to the report.pdf file)

2) Plot showing  standard deviation of the particle defocus per micrograph (micrograph with high deviation are those in with the defocus assigned to the different particles is more different) (This plot is not available in relion)

3) gallery of particles showing defocus and beamtitl

4) Two images related with the beamtilt estimation:

the image with the phase differences from which this estimate was derived: beamtilt_delta-phase_per-pixel.mrc  and the model fitted through it: beamtilt_delta-phase_lin-fit.mrc --scale 3 &

--------------------------------------------------------------

All the plots are created from an auxiliary database that is created by the protocol.

That will be all.

### CODE REVIEW

**Comment by delarosatrevin** [comment]
Hi @rmarabini, first of all, thank a lot for improving the visualization for this protocol.

I have quickly went through the code and I haven't tested yet. But the is a thing that I don't like much. It is the direct use of Sqlite3 library and SQL commands. I think this is not something that we should advertise or promote for other developers to follow...and in the long term, it will be harder to maintain if you don't have you SQL commands fresh. 

Can I kindly request you to change the implementation of this DB using our Set class? This class will use the mapper underneath, that is the whole point of the mapper, to hide some of the SQL details form the developers. I think we have now all tools for indexes, aggregation, etc. I think it should be easy to create a generic Set, that contains Objects...and you can set any attribute to them (Float, Integer, etc). 

Thanks again for your contribution.

File: relion/protocols/protocol_ctf_refinement.py [hunk 1]
```diff
@@ -155,9 +166,11 @@ def _getInputVolumes(self, postStar):
     def refineCtfStep(self):
         args = "--i %s " % self._getPath('input_particles.star')
         args += "--o %s " % self._getExtraPath()
-        postStar = self.inputPostprocess.get()._getExtraPath('postprocess.star')
+        postStar = self.inputPostprocess.get().\
```
Comment (delarosatrevin) [inline relion/protocols/protocol_ctf_refinement.py]: Please, follow PEP8 conventions!!! No \ for breaking lines.

File: relion/protocols/protocol_ctf_refinement.py [hunk 2]
```diff
@@ -155,9 +166,11 @@ def _getInputVolumes(self, postStar):
...
+            _getExtraPath('postprocess.star')
         args += "--f %s " % postStar
-        args += "--m1 %s --m2 %s --mask %s " % relion.convert.getVolumesFromPostprocess(postStar)
+        args += "--m1 %s --m2 %s --mask %s " % \
```
Comment (delarosatrevin) [inline relion/protocols/protocol_ctf_refinement.py]: Same here.

File: relion/protocols/protocol_ctf_refinement.py [hunk 3]
```diff
@@ -180,11 +193,69 @@ def refineCtfStep(self):
...
+            micName = coordinate.getMicName()
+            coordX = coordinate.getX()
+            coordY = coordinate.getY()
+            defocusDiff = p2D - p1D
+            defocus = p2D
+            data.append((micId, micName,
+                         coordX, coordY,
+                         defocusDiff, defocus))
+        try:
+            sql = "INSERT INTO %s  " \
```
Comment (delarosatrevin) [inline relion/protocols/protocol_ctf_refinement.py]: More \ over here.

**Comment by delarosatrevin** [comment]
One more thing, please take a look a the code style, in many places PEP8 is not followed. Mostly using \ for breaking lines, there are other recommended ways to achieve the same.

**Comment by rmarabini** [comment]
Hi @delarosatrevin ,

   I have tried to use the mapper instead of sqlite to store the auxiliary data needed by the mapper. The first part was easy, I created a class called "_Micrograph" -that inherits from EMobject"and another called "_setOfMics" -that inherits from EMset-.Using both classes the information needed by the viewer is saved in a file called analyze.sqlite. Unfortunately, I do not know how to read the file. Each time I try to read it I get the error 

WARNING: Class '_Micrograph' not found in mapper dict. Ignored.  

Can you look at the code (file viewer_ctfrefine.py) and suggest a way to read the file? I have pushed the code to the repository scipion-em-relion branch rm_ctfrefineanalyze. The problematic line is in __init__ line 26:

        self.setOfMic = _SetOfMics(filename=setOfMicFnName)


  Thabks

**Comment by delarosatrevin** [comment]
Hi @rmarabini, thanks a lot for addressing the suggested changes. 
I will try to look at the issue with the Mapper to retrieve the data as soon as possible. Additionally, I find  a bit weird that the classes (and logic) for the viewer are defined in the protocol, it should better placed in the viewer code. Do you mind if I try to move this code to the viewer? (I'm maybe a little renaming as well, since _Micrograph and _SetOfMics do not follow our conventions and can be a bit confusing)

**Comment by rmarabini** [comment]
Hi @delarosatrevin,

   Go ahead, move and rename. ;-)

   Roberto

**Comment by delarosatrevin** [comment]
Hi @rmarabini I fixed the Set issue (basically, you need to provide the classesDict, since new objects not explicitly include in the DataModel are not recognized by the mapper) and finished with some (or a lot) clean-up...mainly moving code to the viewer and avoiding using so much global variables. I also set the defocus plot as 2D, 3D seemed weird to me...still the lines are commented, so it is easy to revert back to 3D. Check that everything is working and if you want to add something else...then maybe @azazellochg , could you take a final look before merging?

**Comment by rmarabini** [comment]
Hi    @delarosatrevin <https://github.com/delarosatrevin>,

    Thanks for solving the DataModel issue, I will go over the code as soon
as a can.

**Comment by azazellochg** [comment]
I'll have a look when I can as well.

**Comment by delarosatrevin** [comment]
Thanks @azazellochg , maybe best wait until all changes from @rmarabini are done. ;)

**Comment by rmarabini** [comment]
Hi @delarosatrevin,

  I am reviewing this pull request. When executing analyze I get the error

00048:   Protocol failed: 'ProtRelionCtfRefinement' object has no attribute '_createMicObject'

did you forget to upload the file in which ProtRelionCtfRefinement is defined or should I look into a different branch?

**Comment by delarosatrevin** [comment]
Fixed @rmarabini , I forgot to remove that line from the protocol. Now it should work.

**Comment by rmarabini** [comment]
hi @delarosatrevin,

  I would like to introduce a few modification but before touching the code let us agree on what should be done:

1) the computation of infoset should be done in the protocol not in the viewer. My test data set size is 23000 particles and analyze needed 15 sec to create the infoSet.  15 sec waiting for the window is a long time and data sets ten times larger are not uncommon. On the other hand ctf refine take 2 hours for this data set so a few extra second will not be noticed.

2) The dimensions of the plot containing the particle defocus change from micrograph to micrograph. I thick we should keep the size of the axis fixed. Ideally the maximum value for x and y axis should be the micrograph size. I do not know how to get this size (do you?) but we can at least use as maximum X and Y the values for the particles with large X an Y coordinates

3) there is a estrange relation between the plots with the particle position and the plot with the stdev.  If  I open the plots in this order (first particle position then stdev) when I press the arrow key to move to the next micrograph the content of the particle plot is showed  in the stdev plot (both windows get the same type of plots). This problem does not appear if we show first the stdev plot and then the particle plot. Therefore, I propose, in analyze form,  to merge the "show defocus" and "show defocus variance per micrograph" buttons. That is, open both plots with a single click and open them in the "right" order. Actually, this modification make sense since the way to look at the plots is not go one by one but search for those with higher stdev and then visualize that particular micrograph.

**Comment by delarosatrevin** [comment]
HI @rmarabini, 

1) I understand the problem here...I still don't like much generating the visualization data in the protocol code (I think we should keep separated the computation logic from the visualization logic). Nonetheless, I understand the issue and we should address it. @azazellochg any comment here?

2) Yes, we could try to get the micrographs  size here (@rmarabini , maybe you should go back to first year of Scipion course ;) 

3) I'm a bit lost about the logic of the plot, if you think it make more sense in this way, I'm fine with it.

**Comment by azazellochg** [comment]
Hi @rmarabini @delarosatrevin ! @rmarabini I like the new plots! Thank you for doing this.
I think I do not mind the last two points. About moving computation from viewer to protocol I think it's fine to make user wait a few minutes as long as:
1) scipion project window is not blocked / frozen
2) relion users are normally wait a lot of time when displaying big datasets / star files in relion_display, so they won't be surprised
3) there should be indication in stdout console or somewhere that it's computing stuff, smth like printing "3/120 mics processed" or similar would do

**Comment by rmarabini** [comment]
Hi @azazellochg  @delarosatrevin,

   I will try to convince you once again about computing the infoset  in the protocol. I understand that makes sense to keep independent the logic of the program (the protocol) from the visualization of the results (analyze). So I willing to put all the code related with visualization in the viewer file. But, even if this general rue is slightly violated,  I do not see why the function that creates the infoset cannot be called from the protocol if this increases efficiency. The advantage is clear, the user  gets an intermediate visualization vs waiting a few minutes.
May be scipion users are used to wait for visualization but we should try to avoid it.

**Comment by delarosatrevin** [comment]
Hi @rmarabini, 
After your "convincing" request, we agreed to move the computation to the protocol. 

1) I already addressed these changes in my last commits. a) moved classes to relion.objects b) Computed info in a new protocol step c) Both protocol and viewer use the data objects from relion.objects.

2) I was taking a look at the micrograph size issue, but it is not easy to achieve at this point, where there is not direct input micrographs.

So, go ahead with your points 2 (maybe max of all coordinates, expensive?) and point 3.

**Comment by rmarabini** [comment]
OK guys,

    I surrender. I will implement point 2 and 3.

                     Roberto

**Comment by delarosatrevin** [comment]
@rmarabini Why do you "surrender"? We have accepted your proposal, and even better, I have implemented point 1 for you !!!

**Comment by rmarabini** [comment]
Oh, I misunderstood you.

    Roberto

**Comment by rmarabini** [comment]
OK  i @azazellochg @delarosatrevin, this is ready to go.

I have:

1)  merged both plots in a single window
2) fixed some problems when the micrographs IDs are not consecutive
3) link both plots so when you click a particular micrograph in tle stdev plot the corresponding microgrphs is shown (and the other way around)
4) compute maxX and maxY values for the micrograph plot.

   hope you enjoy it

**Comment by azazellochg** [comment]
I'm looking at it now.

**Pull request merged**